### DISCUSSION TITLE
Routr on kubernetes

### DISCUSSION THREAD

**Discussion by KjartanRS** [fonoster/yaps#474] [Q&A]
Greetings,
Routr looks like an interesting project and software for use in my company's SIP platform, so I am interesting in doing a proof of concept for my needs - Using yaml files for declarative configuration is of high interest. 
However, we are interested in deploying this in a Kubernetes cluster.
I am not sure how to accomplish this. Looking through your official documentation/git repo, it seems to me like you have tried Kubernetes setup, but it is not yet finished. 

I did find an helm repo on artifact hub, however - this is not updated since April 2021.
Do you have any tips and tricks for getting Routr working in a kubernetes environment?

Also, do you have any experience with Routr in a dev ops environment? (for example using argo CD to qucikly publish new configuration)

**Comment by KjartanRS** [comment]
My idea is to use Routr in a "mobile office" solution. One of my scenarios is that I am in need of continuously updating the Routr configuration (for example the numbers.yml), with maybe several changes every minute.

I have thought about pushing the repo to my local Gitlab instance -> publishing changes to Kubernetes through Argo CD. However, I do believe this will not work, since I am in need of building the image for each change done through the Gitlab repo (in other words; running a Gitlab Pipeline for each time the configuration is changing, which is a time consuming job) 

Alternatively having an app server that push .yml files into Kubernetes for each needed change, but with this solution then I believe Routr will revert back to the "default" image that was originally installed on the k8s cluster if one of the Routr container/pods restarts/re-deploys.

I might need to use the Restapi, but I am not sure if it will work in a Kubernetes environment (through an ingress controller)
https://routr.io/docs/api/overview/

Thank you for your time and please let me know if you have any questions.

**Reply by psanders** [reply]
Hi @TelcoBelco 

Awesome use case! Few things:

I recommend trying first with [v1](https://github.com/fonoster/routr/tree/v1), which is more mature regarding K8s deployments. The [helm](https://github.com/fonoster/routr/tree/v1/.helm) will do the trick.

I will be happy to support any effort to keep the artifact hub up to date, but it needs to be a community effort.

The chart exposes the RestAPI as a service, and you can expose it as any other service.

**Comment by KjartanRS** [comment]
Thank you for the feedback, @psanders.
I have successfully setup the Routr API-instance in an Azure Kubernetes Service enviroment.
Setting up a DNS record, the load balancer, ingress and route. In my route I had to define the; metadata.annotations.nginx.ingress.kubernetes.io/backend-protocol: HTTPS (since the nginx sent http upstream and therefore the Routr server failed)

Going forward I of course need the tcp/udp SIP integration.
I have enabled the tcpsip service through the value.yaml (helm chart) and set it up with my loadbalancer and ingress+route, similar to the working API service.
I have also installed sngrep on the pod, however - I do not see any SIP traffic on my pod.
On my nginx ingress controller, I also do not see successful traffic going to pod:5060 service. The logs report of the following:
`upstream prematurely closed connection while reading response heade ││ r from upstream, client: redacted, server: _, request: "HEAD / HTTP/1.1", upstream: "http://redacted:5060 ││ /",`
Please note the port, 5060 that is defined in the siptcp service, in the upstream service.
I have tried with and without the above mentioned "annotations" to HTTPS on the siptcp & sipudp ingress route.

Any tips in troubleshooting the sip tcp setup on my AKS cluster?

Another, small, thing; 
I wish to setup Routr in "static mode" on the gateways. Ref the documentation, I am not able to set this up through the API service outputting the following message: "message": "$[0].spec.port: string found, integer expected"
Simply removing the "port" will work, but then I am not able to define what port is used in static mode(?)
Are you familiar with any method of updating the port value through the API? (doing this in the gateways.yaml file is not optimal, since I want a most possible git ops setup. So API is my preferred way.). This might be a bug or not a feature through the API, I am using the v1 version, as you earlier recommended.
![image](https://user-images.githubusercontent.com/121808068/215442200-fadc74de-60bb-4720-a970-2c19d385dfd2.png)
https://routr.io/docs/configuration/gateways/#example-if-gateway-using-static-mode

**Reply by psanders** [reply]
I apologize for the delay in response. It looks like this conversation may have gone unnoticed.

Regarding the issue you mentioned, it sounds like there may be a problem with deserializing the number correctly. I suggest replacing the string "5060" with the number 5060 which may fix the issue.

If you have any additional information or if there's anything else I can assist with, please let me know. I'd be happy to help!

**Comment by psanders** [comment]
I will close this conversation to maintain organization. However, if you need further assistance in the future, feel free to re-open it.
